src/intrinsics/ecma262/Object.js

```diff
@@ -101,36 +129,67 @@ export default function(realm: Realm): NativeFunctionValue {
         }
         keys = frm.$OwnPropertyKeys();
 
-        if (to_must_be_partial) {
-          if (to instanceof AbstractObjectValue && to.values.isTop()) {
-            // We don't know which objects to make partial and making all objects partial is failure in itself
-            AbstractValue.reportIntrospectionError(to);
-            throw new FatalError();
-          } else {
-            // if to has properties, we better remove them because after the temporal call to Object.assign we don't know their values anymore
-            if (to.hasStringOrSymbolProperties()) {
-              // preserve them in a snapshot and add the snapshot to the sources
-              delayedSources.push(to.getSnapshot({ removeProperties: true }));
-            }
+        handleSnapshot(frm, frm_was_partial);
+
+        // c. Repeat for each element nextKey of keys in List order,
+        invariant(frm, "from required");
+        invariant(keys, "keys required");
+        copyKeys(keys, frm, to);
+      };
+
+      const tryAndApplySourceOrRecover = nextSource => {
+        let effects;
+        let savedSuppressDiagnostics = realm.suppressDiagnostics;
+        try {
+          realm.suppressDiagnostics = true;
+          effects = realm.evaluateForEffects(
+            () => {
+              applySource(nextSource);
+              return realm.intrinsics.undefined;
+            },
+            undefined,
+            "tryAndApplySourceOrRecover"
+          );
+        } catch (e) {
+          let frm = To.ToObject(realm, nextSource);
+          let validFrom = frm.mightNotBeHavocedObject();
```

Comment:
Why is it valid? !frm.mightNotBeHavocedObject() tells you for sure that frm is havoced, while !frm.mightBeHavocedObject() tells you for sure that frm is not havoced. Without the negation you don't really know anything.

Reply:
I'm not even sure why @NTillmann even added the "might be" logic then. If we do this, we never once enter this new bail out logic in every test we have and in all our internal tests. In fact, it's actually impossible for `validFrom` to ever be true here.

Reply:
Done